docs/architecture/adr-071-bank-v2.md

```diff
@@ -0,0 +1,183 @@
+
+**Legacy Support**: A legacy wrapper will be implemented to ensure compatibility with about 90% of existing functions. This measure will facilitate a smooth transition while keeping older systems functional.
+
+**Callback Functions**: We propose to integrate callback functions directly to the x/bank module, allowing for customisable behaviour during minting, burning and transferring operations without complicating statemanagement. It also helps overwrite the default accounting mechanism by allowing to overwrite the balance method (this is particularly useful for rebasing tokens).
+
+Implementation of Callback Functions:
+
+```go
+type BankKeeper struct {
+    OnMint func(ctx context.Context, addr []byte, amount sdk.Coin) error
+    OnBurn func(ctx context.Context, addr []byte, amount sdk.Coin) error
+    OnTransfer func(ctx context.Context, fromAddr, toAddr []byte, amount sdk.Coin) error
+    OnBalanceOf func(ctx context.Context, addr []byte, denom string) (sdk.Coins, error)
+}
+
+func (k *BankKeeper) MintCoins(ctx context.Context, moduleName string, amt sdk.Coins) error {
+    if k.OnMint != nil {
+        if err := k.OnMint(ctx, moduleName, amt); err != nil {
+            return err
+        }
+    }
+}
+
+func (k *BankKeeper) BurnCoins(ctx context.Context, moduleName string, amt sdk.Coins) error {
+    if k.OnBurn != nil {
+        if err := k.OnBurn(ctx, moduleName, amt); err != nil {
+            return err
+        }
+    }
+}
+
+func (k *BankKeeper) SendCoins(ctx context.Context, fromAddr, toAddr []byte, amt sdk.Coins) error {
+    if k.OnTransfer != nil {
+        if err := k.OnTransfer(ctx, fromAddr, toAddr, amt); err != nil {
+            return err
+        }
+    }
+}
+
+func (k *BankKeeper) GetBalance(ctx context.Context, addr []byte, denom string) (sdk.Coins, error) {
+    if k.OnBalanceOf != nil {
+        coins, err := k.OnBalanceOf(ctx, addr, denom);
+        if err != nil {
+            return nil, err
+        }
+
+        return coins, err
+    }
+}
+```
+
+Messages for Admin Operations:
+
+To facilitate admin operations like `Mint`, `Burn`, and `Transfer`, we can define specific messages that can be invoked by the admin account of each denom. These operations will be internal and not exposed to end clients directly.
+
+```protobuf
+message Mint {
+  string denom = 1;
+  string to = 2;
+  string amount = 3;
+}
+
+message Burn {
+  string denom = 1;
+  string from = 2;
+  string amount = 3;
+}
+
+message Transfer {
+  string denom = 1;
+  string from = 2;
+  string to = 3;
+  string amount = 4;
+}
+```
+
+If admin account defines these callbacks, `x/bank` will call `x/accounts` `MsgExecute` with the `OnTransfer`, `GetBalance` or `GetSupply` callbacks:
```

Comment:
call back on queries would need to also be in the indexing layer right?

Reply:
Not sure quite what you mean. The indexing layer wouldn't index any balances or supplies that aren't managed by x/bank. For those we would need to figure out some indexing solution for the x/accounts implementations.